### DISCUSSION TITLE
Support for STM32F401 non nucleo (i.e. "generic") board?

### DISCUSSION THREAD

**Discussion by TRadigk** [micropython/micropython#13083] [STM32 / Pyboard] [answered]
Hi guys,
long story short, I am currently developing a C# based driver for micropython (file transfer and so via raw repl) and wanted to test with STM32 ports. Since in my projects I am only using RP2040 and ESP32 I thought the cheapest board should do the trick and I got this one:
https://www.berrybase.de/stm32f401-development-board-v3.0#

But I can't find any resources about how and which hex file to program. I could only find information about STM32F4**1**1 or that in general micropython's minimal requirements are 256 KB flash and 64 KB Ram, which are met with this board.
I also can only flash via st link / SWD at the moment.

**Comment by robert-hh** [comment]
There are two boards with an STM32F401 with pre-built firmware, the Espruino Pico and Nucleo F401RE. These firmware could work, bit maybe problems in the Pin name matching. I would take that as starting point.

**Reply by TRadigk** [reply]
Hi robert,
This would mean that I have to set up C development and make a new firmware myself and get to know how to change the pins, correct? If there is no other way, then I will plan to that or maybe never do that, because I am not really a C Programmer (quite high effort for very little outcome, also I have never built any C project). 
If you could confirm, that this is the way it is, I will mark your post as answer for this discussion.

**Reply by robert-hh** [reply, marked as answer]
You could just use one of the existing firmware files and try them. Otherwise: yes, that's what you have to do.

You can start by copying the directory for one of the other F401 boards. Adapting the pins is easy. In the respective board directory there is a pins.csv file, which maps boards names to CPU names. For the Nucleo board, board and cpu name are identical. The other files which may have to be changed is the mpconfigboard.h,which is more tricky and requires knowing the hardware details. The board looks like the WEACT Blackpill board with the stm32F401. More information e.g. here: https://stm32-base.org/boards/STM32F401CCU6-WeAct-Black-Pill-V1.2.html.
Alternatively you can get a board for which pre-built firmware exists. Common favorites are the RPi Pico or one of the ESP32 variants.
